libcore/FileOperations.vala

```diff
@@ -0,0 +1,230 @@
+namespace Marlin.FileOperations {
+    public class CommonJob {
+        protected unowned Gtk.Window? parent_window;
+        protected uint inhibit_cookie;
+        protected GLib.Cancellable? cancellable;
+        protected PF.Progress.Info progress;
+        protected Marlin.UndoActionData? undo_redo_data;
+        protected CommonJob (Gtk.Window? parent_window = null) {
+            this.parent_window = parent_window;
+            inhibit_cookie = 0;
+            progress = new PF.Progress.Info ();
+            cancellable = progress.get_cancellable ();
+            undo_redo_data = null;
+        }
+
+        ~CommonJob () {
+            progress.finish ();
+            uninhibit_power_manager ();
+            if (undo_redo_data != null) {
+                Marlin.UndoManager.instance ().add_action ((owned) undo_redo_data);
+            }
+        }
+
+        protected void inhibit_power_manager (string message) {
+            weak Gtk.Application app = (Gtk.Application) GLib.Application.get_default ();
+            inhibit_cookie = app.inhibit (parent_window, Gtk.ApplicationInhibitFlags.LOGOUT | Gtk.ApplicationInhibitFlags.SUSPEND, message);
+        }
+
+        private void uninhibit_power_manager () {
+            if (inhibit_cookie != 0) {
+                return;
+            }
+
+            ((Gtk.Application) GLib.Application.get_default ()).uninhibit (inhibit_cookie);
+            inhibit_cookie = 0;
+        }
+
+        protected bool aborted () {
+            return cancellable.is_cancelled ();
+        }
+    }
+
+    public class EmptyTrashJob : CommonJob {
+        GLib.List<GLib.File> trash_dirs;
+
+        public EmptyTrashJob (Gtk.Window? parent_window = null, owned GLib.List<GLib.File>? trash_dirs = null) {
+            base (parent_window);
+            if (trash_dirs != null) {
+                this.trash_dirs = (owned) trash_dirs;
+            } else {
+                this.trash_dirs = new GLib.List<GLib.File> ();
+                this.trash_dirs.prepend (GLib.File.new_for_uri ("trash:"));
+            }
+        }
+
+        /* Only called if confirmation known to be required - do not second guess */
+        private bool confirm_empty_trash () {
```

Comment:
Need to implement "DESTRUCTIVE_ACTION" styling on the EMPTY_TRASH button.  There is already a dialog that does this (StockDIalogs.vala run_simple_file_operation_dialog ()), although that will need to be modified/split to work sanely with FIleOperations.vala

Reply:
add_button returns the actual GtkButton widget so I just added the DESTRUCTIVE ACTION Class there :)